feat: impl subtract for primitive_scalar

doki23 opened the pull request:
part of #1591

doki23 committed:
impl subtract for primitive_scalar

gatesn commented:
Hey @doki23, are these PRs ready for review? Mark them non-draft when ready and we can take a look

doki23 commented:
> Hey @doki23, are these PRs ready for review? Mark them non-draft when ready and we can take a look

Yes, I had a travel last days:), I'll change its state, PTAL

doki23 commented:
Maybe we could impl `std::ops::Sub` for it.

gatesn commented:
Can you explain a bit about how this is going to be used? I'm wondering whether these operations should be implemented on `Scalar`, `PrimitiveScalar`, or `PValue`.

doki23 commented:
> Can you explain a bit about how this is going to be used? I'm wondering whether these operations should be implemented on `Scalar`, `PrimitiveScalar`, or `PValue`.

If we implement `fn subtract_scalar(&self, array: &ArrayData, to_subtract: &Scalar) -> VortexResult<ArrayData>` for ConstantEncoding, we can get the result scalar by `array.scalar().as_primitive() - to_subtract.as_primitive()`.

doki23 commented:
Hey @gatesn , please take a look and let's move it forward.

gatesn commented on vortex-scalar/src/primitive.rs:
```diff
@@ -103,6 +105,27 @@ impl<'a> PrimitiveScalar<'a> {
         }
     }
+
+    pub fn subtract(&self, other: &PrimitiveScalar) -> VortexResult<Self> {
```
Could you call this `checked_sub(&self, other: &PrimitiveScalar) -> VortexResult<Option<Self>>` and then use num-traits CheckedSub to ensure the subtraction doesn't overflow. Otherwise this function can panic.

Otherwise looks good to me!

doki23 replied:
I'm not sure the float checked_sub is correct, please review.

doki23 committed:
checked_sub

gatesn commented on vortex-scalar/src/primitive.rs:
```diff
@@ -103,6 +108,46 @@ impl<'a> PrimitiveScalar<'a> {
+                }
+            })
+        } else {
+            match_each_float_ptype!(self.ptype, |$T| {
```
I would just do regular subtraction for floats.  There's not exactly overflow behavior

doki23 replied:
I think it's ok now

doki23 committed:
regular sub for float values

gatesn commented:
Sorry to be a pain again! @danking merged a change to the macro to let you implement different code blocks for ints and floats, e.g. https://github.com/spiraldb/vortex/pull/1640/files#diff-8f1cbe207a4076299ad7913ac7d7d0db5fd2e99d1a689f984fcedfd1b91208faR308-R329

And also actually that PR added arithmetic operations to arrays. Not sure if that's sufficient for you?

danking committed:
Merge remote-tracking branch 'origin/develop' into HEAD

danking commented:
I merged and switched to the new macro for you @doki23 .

My general inclination is to use the (just added; sorry, we were working in parallel!) `checked_numeric_operator` method I just added, but there's a key difference between my implementation and yours. You avoid allocating or cloning a `DType` for each operation.

In the use cases I had (implementing BinaryNumericFn for ConstantArray and SparseArray), I always needed a Scalar anyway. But I think your approach seems better. Let me rework checked_numeric_operator to return a `PrimitiveScalar<'a>`. Then I'll push that to this branch and use it as the implementation of checked_sub. There's a bit more overhead (a match on the operator), but I think this will strike a good balance between maintainability while still avoiding allocating the DType if you don't want it.

I'll plan to have this pushed later today.

danking committed:
1. `checked_numeric_operator` returns a PrimitiveScalar (avoiding a DType allocation)
2. `checked_sub` delegates to `checked_numeric_operator`

As a consequence of (2), `checked_sub` now returns an `None` when there is an underflow/overflow
rather than returning a Null PrimitiveScalar (the return type is now
`VortexResult<Option<PrimitiveScalar<'a>>>`).

Moreover, `std::ops::Sub` now returns an error on underflow/overflow rather than a Null
PrimitiveScalar.

danking commented:
Okay, here's what I did:

1. checked_numeric_operator returns a PrimitiveScalar avoiding a DType allocation

2. `checked_sub` delegates to `checked_numeric_operator`

As a consequence of (2), `checked_sub` now returns a `None` when there is an underflow/overflow
rather than returning a Null PrimitiveScalar (the return type is now
`VortexResult<Option<PrimitiveScalar<'a>>>`).

Moreover, `std::ops::Sub` now returns an error on underflow/overflow rather than a Null
PrimitiveScalar.

danking commented:
@doki23 does this suit your needs?

doki23 commented:
> @doki23 does this suit your needs?

Thanks, I'll have a look.

doki23 commented:
Hey @danking , thanks for your excellent work! I have only one suggestion — I prefer using a `into` method to convert `PrimitiveScalar` to `Scalar` rather than using `to_scalar`. If you also agree, I'll make this change.

doki23 committed:
implement From<PrimitiveScalar> for Scalar

danking commented:
Sounds good. I’ll approve with that change

doki23 commented:
> Sounds good. I’ll approve with that change

I've done, PTAL @danking
